crypto/initthread.c

```diff
@@ -424,16 +433,49 @@ int ossl_init_thread_start(const void *index, void *arg,
         return 0;
 
     hand->handfn = handfn;
+    hand->prio = prio;
     hand->arg = arg;
 #ifndef FIPS_MODULE
     hand->index = index;
 #endif
-    hand->next = *hands;
-    *hands = hand;
+
+    /*
+     * We need to do an ordered insert here, with entries at a higher
+     * priority value appearing later in the list
+     */
+    prev = NULL;
+    idx = *hands;
+    while (idx != NULL) {
+        entry_count++;
+        if (idx->prio <= hand->prio) {
+            prev = idx;
+            idx = idx->next;
+        } else {
+            break;
+        }
+    }
+    if (entry_count == 0) {
+        hand->next = *hands;
+        *hands = hand;
+    } else {
+        if (prev == NULL) {
```

Comment:
Is there a different scenario than `prev == NULL && entry_count == 0` to check for this here?